inc/Store/DataSource/DataSourceConfigManager.php

```diff
@@ -0,0 +1,248 @@
+		return array_map(
+			function ( array $config ) {
+				return array_merge( $config, [ 'config_source' => self::CONFIG_SOURCE_CODE ] );
+			},
+			ConfigStore::get_data_sources_as_array()
+		);
+	}
+
+	/**
+	 * Quick and dirty de-duplication of data sources. If the data source does
+	 * not have a UUID (because it is registered in code), we generate an
+	 * identifier based on the display name and service name.
+	 */
+	private static function de_duplicate_configs( array $configs ): array {
+		return array_values( array_reduce(
+			$configs,
+			function ( array $acc, array $item ) {
+				$identifier = $item['uuid'] ?? md5(
+					sprintf( '%s_%s', $item['service_config']['display_name'], $item['service'] )
+				);
+				$acc[ $identifier ] = $item;
+				return $acc;
+			},
+			[]
+		) );
+	}
+
+	/**
+	 * Get all data sources from all origins with optional filters.
+	 * 
+	 * Supported filters:
+	 * - service: Filter by service name (e.g. 'airtable', 'google-sheets', 'shopify')
+	 * - enable_blocks: Filter by blocks enabled status (false matches with null/false and true matches with true)
+	 * 
+	 * Any unsupported filter keys will be ignored.
+	 * 
+	 * @param array{
+	 *   service?: string,
+	 *   enable_blocks?: bool
+	 * } $filters Optional filters to apply to the results.
+	 * @return array<array{
+	 *   uuid?: string,
+	 *   service: string,
+	 *   service_config: array<string, mixed>,
+	 *   config_source: string,
+	 *   __metadata?: array{
+	 *     created_at: string,
+	 *     updated_at: string
+	 *   }
+	 * }>
+	 */
+	public static function get_all( array $filters = [] ): array {
+		$code_configured = self::get_all_from_code();
+		$constant_configured = self::get_all_from_constant();
+		$storage_configured = self::get_all_from_storage();
+
+		/**
+		 * De-duplicate configs.
+		 * 
+		 * Precedence (lowest to highest):
+		 * - Code-configured data sources
+		 * - Constant-configured data sources
+		 * - Storage-configured data sources
+		 */
+		$configs = self::de_duplicate_configs(
+			array_merge( $code_configured, $constant_configured, $storage_configured )
+		);
+
+		return self::apply_filters( $configs, $filters );
+	}
+
+	/**
+	 * Apply filters to an array of configs.
+	 *
+	 * @param array $configs The configs to filter.
+	 * @param array{
+	 *   service?: string,
+	 *   enable_blocks?: bool
+	 * } $filters The filters to apply.
+	 * @return array The filtered configs.
+	 */
+	private static function apply_filters( array $configs, array $filters ): array {
+		if ( empty( $filters ) ) {
+			return $configs;
+		}
+
+		return array_filter( $configs, function ( array $config ) use ( $filters ): bool {
+			foreach ( $filters as $key => $value ) {
+				$passes_filter = match ( $key ) {
+					'service' => $config['service'] === $value,
+					'enable_blocks' => ( $config['service_config']['enable_blocks'] ?? false ) === $value,
+					default => true,
+				};
+
+				if ( ! $passes_filter ) {
+					return false;
+				}
```

Comment:
I don't see how`$passes_filter` can ever be `false` given the final default expression. Maybe you want the default expression to resolve to `null` or some constant that represents an empty set?

Also, can you add type hints (param and return) to the `match`? They are new to me.

Reply:
Ignore the first part, but still curious about types.